src/i_3dsound.c

```diff
@@ -0,0 +1,312 @@
+//
+// Copyright(C) 2023 ceski
+//
+// This program is free software; you can redistribute it and/or
+// modify it under the terms of the GNU General Public License
+// as published by the Free Software Foundation; either version 2
+// of the License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// DESCRIPTION:
+//      System interface for 3D sound.
+//
+
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "doomstat.h"
+#include "i_oalsound.h"
+#include "r_data.h"
+#include "r_main.h"
+
+#define VIEWDIST (ORIGWIDTH / 2) // Distance from projection plane to POV.
+#define FIXED_TO_ALFLOAT(x) ((ALfloat)(((double)x) / FRACUNIT))
+
+typedef struct oal_listener_params_s
+{
+    ALfloat position[3];
+    ALfloat velocity[3];
+    ALfloat orientation[6];
+} oal_listener_params_t;
+
+typedef struct oal_source_params_s
+{
+    ALfloat position[3];
+    ALfloat velocity[3];
+    ALfloat direction[3];
+    boolean point_source;
+} oal_source_params_t;
+
+static boolean I_3D_InitSound(void)
+{
+    return I_OAL_InitSound(true);
+}
+
+static boolean I_3D_ReinitSound(void)
+{
+    return I_OAL_ReinitSound(true);
+}
+
+static void CalcPitchAngle(const player_t *player, angle_t *pitch)
+{
+    fixed_t lookdir, slope;
+
+    // Same as R_SetupFrame() in r_main.c.
+    lookdir = player->lookdir / MLOOKUNIT + player->recoilpitch;
+
+    if (lookdir == 0)
+    {
+        *pitch = 0;
+        return;
+    }
+
+    if (lookdir > LOOKDIRMAX)
+        lookdir = LOOKDIRMAX;
+    else if (lookdir < -LOOKDIRMAX)
+        lookdir = -LOOKDIRMAX;
+
+    slope = (lookdir << FRACBITS) / VIEWDIST;
+    if (slope < 0)
+    {
+        *pitch = (ANGLE_MAX - tantoangle[-slope >> DBITS]) >> ANGLETOFINESHIFT;
+    }
+    else
+    {
+        *pitch = tantoangle[slope >> DBITS] >> ANGLETOFINESHIFT;
+    }
+}
+
+static void CalcListenerParams(const mobj_t *listener, oal_listener_params_t *lis)
+{
+    const player_t *player = listener->player;
+    const angle_t yaw = listener->angle >> ANGLETOFINESHIFT;
+    angle_t pitch;
+
+    lis->position[0] = FIXED_TO_ALFLOAT(listener->x);
+    lis->position[1] = FIXED_TO_ALFLOAT(player->viewz);
+    lis->position[2] = FIXED_TO_ALFLOAT(-listener->y);
+
+    lis->velocity[0] = FIXED_TO_ALFLOAT(listener->momx) * TICRATE;
```

Comment:
Do we even need to set these if `snd_doppler == 0`?

Reply:
Fixed by https://github.com/fabiangreffrath/woof/pull/1118/commits/327565a2427ae7687c13f6b48f71d6cd2e33ee89